File: pat-static/src/test/java/com/rectanglel/patstatic/buildutils/TrueTimeDataCacher.java
package com.rectanglel.patstatic.buildutils;

import com.rectanglel.patstatic.model.PatApiService;
import com.rectanglel.patstatic.model.PatApiServiceImpl;
import com.rectanglel.patstatic.routes.BusRoute;

import java.io.File;
import java.util.concurrent.TimeUnit;

/**
 * Cache all routes to a folder
 * <p>
 * One thing I need to watch out for is that the
 * Created by epicstar on 3/5/17.
 * @author Jeremy Jao
 */

public class TrueTimeDataCacher {

    private PatApiService patApiService;
    private File cacheDirectory;

    public TrueTimeDataCacher(String baseUrl, String apiKey, File cacheDirectory) {
        this.cacheDirectory = cacheDirectory;
        patApiService = new PatApiServiceImpl(baseUrl, apiKey, this.cacheDirectory, new StubStaticData(), new StubWifiDataChecker());
    }

    /**
     * Cache routes if not present.
     */
    public void cacheAllRoutes() {
        patApiService.getRoutes()
                .toObservable()
                .flatMap(routes ->
                        io.reactivex.Observable.fromIterable(routes)
                                .skipWhile(route -> {
                                    String routeNumber = route.getRouteNumber();
                                    return new File(cacheDirectory, String.format("lineinfo/%s.json", routeNumber)).exists();
                                })
                                .zipWith(io.reactivex.Observable.interval(0, 500, TimeUnit.MILLISECONDS), (stuff, aLong) -> stuff)
                )
                .buffer(8)
                .zipWith(io.reactivex.Observable.interval(0, 5, TimeUnit.SECONDS), (stuff, aLong) -> stuff)
                .flatMapIterable(routes1 -> routes1)
                .map(BusRoute::getRouteNumber)
                .flatMap(routeNumber -> patApiService
                        .getPatterns(routeNumber)
                        .onErrorResumeNext(io.reactivex.Observable.empty())
                        .doOnNext((pattern) -> System.out.println(String.format("Saving route number: %s", routeNumber)))
                );
    }

    public static void main(String[] args) {
        String baseUrl = args[0];
        String patApiKey = args[1];
        File cacheDirectory = new File(args[2]);
        TrueTimeDataCacher cacher = new TrueTimeDataCacher(baseUrl, patApiKey, cacheDirectory);
        cacher.cacheAllRoutes();
    }
}

File: pat-static/src/test/java/com/rectanglel/patstatic/buildutils/TrueTimeDataCacher.kt
package com.rectanglel.patstatic.buildutils

import com.rectanglel.patstatic.model.PatApiService
import com.rectanglel.patstatic.model.PatApiServiceImpl
import com.rectanglel.patstatic.routes.BusRoute
import io.reactivex.Flowable
import io.reactivex.Observable
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Cache all routes to a folder
 *
 *
 * One thing I need to watch out for is that the
 * Created by epicstar on 3/5/17.
 * @author Jeremy Jao
 */
class TrueTimeDataCacher(apiKey: String, private val cacheDirectory: File) {
    private val patApiService: PatApiService

    /**
     * Cache routes if not present.
     */
    fun cacheAllRoutes() {
        patApiService.routes
                .toObservable()
                .flatMap { routes: List<BusRoute> ->
                    Observable.fromIterable(routes)
                            .skipWhile { route: BusRoute ->
                                val routeNumber = route.routeNumber
                                File(cacheDirectory, "lineinfo/$routeNumber.json").exists()
                            }
                            .zipWith(Observable.interval(0, 500, TimeUnit.MILLISECONDS)
                            ) { stuff: BusRoute, _: Long -> stuff }
                }
                .buffer(8)
                .zipWith(Observable.interval(0, 5, TimeUnit.SECONDS)
                ) { stuff, _: Long -> stuff }
                .flatMapIterable { it }
                .map(BusRoute::routeNumber)
                .flatMap { routeNumber: String ->
                    patApiService
                            .getPatterns(routeNumber)
                            .onErrorResumeNext(Flowable.empty())
                            .toObservable()
                            .doOnNext { println(String.format("Saving route number: %s", routeNumber)) }
                }.toFuture().get()
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val patApiKey = args[0]
            val cacheDirectory = File(args[1])
            val cacher = TrueTimeDataCacher(patApiKey, cacheDirectory)
            cacher.cacheAllRoutes()
        }
    }

    init {
        patApiService = PatApiServiceImpl(apiKey, cacheDirectory, StubStaticData(), StubWifiDataChecker())
    }

}